Upstream SenseHAT patches

6by9 opened the pull request:
From #6367, noting that there are upstream patches proposed for the SenseHAT.

The joystick driver was merged, but not these ones. PR for reference and test.

Commit:
drivers/mfd: sensehat: Add Raspberry Pi Sense HAT to simple_mfd_i2c

This patch adds the compatible string for the Sense HAT device to
the list of compatible strings in the simple_mfd_i2c driver so that
it can match against the device and load its children and their drivers

Co-developed-by: Mwesigwa Guma <mguma@redhat.com>
Signed-off-by: Mwesigwa Guma <mguma@redhat.com>
Co-developed-by: Joel Savitz <jsavitz@redhat.com>
Signed-off-by: Joel Savitz <jsavitz@redhat.com>
Signed-off-by: Charles Mirabile <cmirabil@redhat.com>

6by9 committed:
dtoverlays: Convert SenseHAT overlays to use MFD and upstream drivers

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
drivers: Remove downstream SenseHAT core and joystick drivers

Parts of a SenseHAT driver have been submitted upstream using the
simple-i2c-mfd framework. The joystick driver has been merged.

It's been noted that there are several issues with the downstream
joystick and core drivers, so remove these in favour of the upstream
approach, and fix up the FB driver to match.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
defconfigs: Update to build new SenseHAT joystick driver

Dropped downstream JOYSTICK_RPISENSE and use the upstream
JOYSTICK_SENSEHAT instead.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
Input: sensehat-joystick : Revert to downstream keymap

Upstream chose to use BTN_DPAD_* and BTN_SELECT, whilst downstream
had used KEY_*.

Revert to the downstream map to avoid any regressions.
(Ideally this would be read from DT)

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
dtoverlays: Add Sense Hat to hat_map

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

paulober commented:
@6by9 Thanks

I was actually doing exactly the same changes (about 80% I would say) but didn't finished getting the FB driver running again after removing the MFD core driver. Sorry for not communicating.
May I ask, why you did not add `depends on FB && I2C` and `select MFD_SIMPLE_MFD_I2C` instead of the custom MFD driver?

Also do you think it can be a good idea to add
```diff
+config MFD_RASPBERRYPI_SENSE_HAT
+       tristate "Raspberry Pi Sense HAT MFD"
+       depends on I2C
+       select MFD_SIMPLE_MFD_I2C
+       help
+         This module supports the LED Matrix controller found on the Raspberry Pi
+         Sense HAT v1 and v2 boards, and the joystick on the Sense HAT.
+         (Functionally it relies on MFD_SIMPLE_MFD_I2C to provide the framework
+         that loads the child drivers).
```
to `drivers/mfd/Kconfig` like the POE HAT MFD record in there?

pelwell commented:
> This does still need the UUID of the v1 sense HAT adding to hat_map.dts.

What, like this?:
```diff
diff --git a/arch/arm/boot/dts/overlays/hat_map.dts b/arch/arm/boot/dts/overlays/hat_map.dts
index be38b817022f6..ae5d9925c5f4e 100644
--- a/arch/arm/boot/dts/overlays/hat_map.dts
+++ b/arch/arm/boot/dts/overlays/hat_map.dts
@@ -96,6 +96,11 @@ recalbox-rgbdual {
                overlay = "recalboxrgbdual";
        };
 
+       sensehat-v1 {
+               uuid = [ 5d960035 8e87 428f 95d8 59852d697754 ];
+               overlay = "rpi-sense";
+       };
+
        sensehat-v2 {
                uuid = [ 1aa9c428 72eb 48da 9306 8c3706ed3653 ];
                overlay = "rpi-sense-v2";
```
That does seem to do the right thing:
```
  6.43 Read /overlays/hat_map.dtb bytes     1921 hnd 0xc441
  6.66 PCIEx1: PWR 0 DET_WAKE 0
  6.81 Read /config.txt bytes     1487 hnd 0xc
  6.91 Read /overlays/rpi-sense.dtbo bytes     1082 hnd 0xc4f4
  6.05 MESS:00:00:06.405349:0: Loaded overlay 'rpi-sense'
```

6by9 commented:
> May I ask, why you did not add `depends on FB && I2C` and `select MFD_SIMPLE_MFD_I2C` instead of the custom MFD driver?

An oversight.
 
> Also do you think it can be a good idea to add
> 
> ```diff
> +config MFD_RASPBERRYPI_SENSE_HAT
> +       tristate "Raspberry Pi Sense HAT MFD"
> +       depends on I2C
> +       select MFD_SIMPLE_MFD_I2C
> +       help
> +         This module supports the LED Matrix controller found on the Raspberry Pi
> +         Sense HAT v1 and v2 boards, and the joystick on the Sense HAT.
> +         (Functionally it relies on MFD_SIMPLE_MFD_I2C to provide the framework
> +         that loads the child drivers).
> ```
> 
> to `drivers/mfd/Kconfig` like the POE HAT MFD record in there?

No, it just pollutes the config space as it'll only enable the parent, not the subdevices.
Enabling either the joystick or FB driver will select MFD_SIMPLE_MFD_I2C, so that is the correct route.

6by9 commented:
> > This does still need the UUID of the v1 sense HAT adding to hat_map.dts.
> 
> What, like this?:

Thank you. I only found a V2 to test with.

6by9 commented:
Pi5 not getting any joystick events is due to no GPIO23 not having been configured as an input.
Pi3 is an input by default, and hence works.

I'm slightly surprised that the pinmuxing doesn't automatically become an input when you request interrupts from a GPIO, but I guess you could want to know when something else has changed the state of an output.

@pelwell What's the correct route for configuring pinmuxing on Pi5? Still `brcm,pins` and `brcm,function`?

pelwell commented:
I'll have to look at why it doesn't just work (and if it should), but the RP1 pinctrl driver accepts the old-style `brcm,` declarations (with a compatibility mapping), so it's best to use that unless the overlay has to be Pi 5-specific.

6by9 commented:
Thanks. I've tried adding the old brcm, declarations, but don't appear to be doing anything.

It also feels odd that asking for the interrupt doesn't appear to claim the GPIO or pinctrl. It's quite possible I've just messed up though.

pelwell commented:
Push what you've got so far.

6by9 commented:
Sorry, I'd left the office.

Changes pushed now.

6by9 commented on arch/arm/boot/dts/overlays/rpi-sense-overlay.dts:
```diff
@@ -12,11 +12,23 @@
+				joystick {
+					compatible = "raspberrypi,sensehat-joystick";
+					interrupts = <23 1>;
+					pinctrl-name = "default";
```
Found my error - a missing "s" on "pinctrl-name" here.
I'd copied mcp23017-overlay, which is also wrong.

pelwell replied:
I missed it too

6by9 reviewed (commented):
I now agree with upstream that using the BTN_DPAD_* defines is more sensible. The current buttons are treated as straight keyboard inputs. I wondered why my quick evtest wiggles resulted in the Pi rebooting, but I kept on managing to do "up" "enter", and the previous command in my bash history tended to be "sudo reboot".

6by9 commented:
One slighty odd thing is that whilst the overlay defines all the nodes for the relevant kernel modules for the sensors, however none of those modules are built by the default defconfigs. I'm guessing that the devices are being accessed from userspace via some random I2C library.

pelwell commented:
Yes, sadly,

6by9 committed:
dtoverlays: Correct pinctrl assignment on mcp23017

The pinctrl definition used "pinctrl-name" which is missing
an "s" from the end.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
